airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java

```diff
@@ -322,14 +322,34 @@ public List<CheckedConsumer<JdbcDatabase, Exception>> getCheckOperations(final J
       checkOperations.add(database -> {
         final String userName = config.get("username").asText();
 
+        // Verify if the user has privilege at user level
         final List<JsonNode> userPrivileges = database.queryJsons(connection -> {
           final PreparedStatement ps = connection.prepareStatement("SELECT userepl FROM pg_user WHERE usename = ?");
           ps.setString(1, userName);
-          LOGGER.info("Verifying required privileges for user: {}", userName);
+          LOGGER.info("Verifying required privileges at user level for user: {}", userName);
           return ps;
         }, sourceOperations::rowToJson);
 
-        if (!userPrivileges.get(0).get("userepl").asBoolean()) {
+        // Verify if the user has privilege at role level
+        final List<JsonNode> rolePrivileges = database.queryJsons(connection -> {
+          final PreparedStatement ps = connection.prepareStatement("""
+              WITH RECURSIVE cte AS (
```

Comment:
So looking recursively what roles the assigned roles have is what is needed?

Reply:
Yes, basically if we have:
granted roleA (No replication) to user
granted roleB (No replication) to roleA
granted roleC (Replication) to roleB

The query will output roleC, so we are sure that a subrole has access.
I added the test to cover that case (only one lvl BTW)

Reply:
Adding extra comment in the function